File: AI Challenge/code/AI/simulate.py
from card import Card, cardDeck
from bot import Bot
import random
import time

def playRound(players):
	# Generate a new shuffled deck
	deck = [val for val in cardDeck]
	random.shuffle( deck )

	# Divide the deck up into player's hands
	for i in range(0, 4):
		players[i].ResetHand( deck[i*13 : (i+1)*13] )

	# Setup round info
	i = 0                 # Start the round at the previous winner
	decompose   = 3       # Stop the round after all player's passed in a row
	start_round = True    # Is this the start of the game
	finished    = False   
	beat = []

	# Find the starting player
	startCard = Card('3D')
	for j in range(0, 4):
		if players[j].hand.contains(startCard) == True:
			i = j
			break

	while not finished:
		# Loop around
		if i > 3:
			i = 0

		# All other players passed
		# Reset the card(s) to beat
		if decompose == 0:
			beat = []
			decompose = 3

		# Allow the current player to play
		play = players[i].Play(
			start_round,
			beat,
			[len(player.hand.remain) for player in players]
		)
		start_round = False

		# Did the bot pass?
		if len(play) == 0:
			decompose -= 1
			continue

		# They made a play, reset decomposure
		decompose = 3

		# Check if the move was valid
		if len(beat) > 0:
			if play[0].lessThan(beat[0]):
				raise ValueError("Invalid play: "+str(play[0])+':'+str(play[0].worth)+' is not greater than '+str(beat[0])+':'+str(beat[0].worth))
		beat = play

		# See if the current player just played their last card
		#  Thus, end game
		if len(players[i].hand.remain) < 1:
			finished = True
			break

		# Allow other bots to know what happened
		for player in players:
			player.Observe(beat, 0)
		
		# Allow the next player to make a turn
		i += 1




	#----------------------------
	#  Scoring
	#----------------------------

	# Determine the number of remaining cards
	# Determine the winner of the trick
	cards = []
	for i, player in enumerate(players):
		cards.append(len(player.hand.remain))

	# Calculate scores
	scores = [0, 0, 0, 0]
	for i in range(0, 4):
		# Determine the winner to be the sum of the rest later
		if cards[i] == 0:
			continue

		if cards[i] < 10:
			scores[i] = -1 * cards[i]
		elif cards[i] <= 12:
			scores[i] = -2 * cards[i]
		else:
			scores[i] = -3 * cards[i]

	
	# Calculate winner's score
	for i in range(0, 4):
		if cards[i] == 0:
			# Sum other player's scores as a postive number
			for j in range(0, 4):
				if j == i:
					continue

				scores[i] -= scores[j]
			break


	return scores


def game(players):
	scores = [0, 0, 0, 0]
	finish = False

	# Play 10 rounds
	for _ in range(0, 10):
		if finish:
			break

		out = playRound(players)

		# Sum all scores
		for j, val in enumerate(out):
			scores[j] += val

		# If a player has passed 100 (there may be multiple)
		#  Stop
		for val in scores:
			if val >= 100:
				finish = True
				break

	return scores



"""
Three ANN bots verse one simplex
Breed from the best and previous best bot
"""
def learnByExample():
	players = [
		Bot(),
		Bot(),
		Bot(),
		Bot()
	]
	players[0].simplex = False
	# Keep the best bot in play for the next round
	players[0].best = False
	players[1].best = False
	players[2].best = False
	players[3].best = False
	bestScore   = 0
	bestNetwork = players[0].network
	prevNetwork = players[0].network

	file = open("log.csv", "w+")
	file.write('Elapsed,Generation,Score 1,Score 2,Score 3,Score 4,Win Rate,Weights,Columns,Largest Column,Neurons\n')

	# Estimation info
	start = time.time()

	smoothedWinRate = 0
	winRate = 0
	wins        = 0
	games       = 0
	generation  = 0
	cycles      = 5000
	logInterval = (cycles/100) # 500 = number of data points at the end of training
	score       = [-100,-100,-100,-100]
	tally       = [0,0,0,0]

	def updateLog(rerun, gameCount, winCount):
		if rerun < 2:
			print(' Est: Unknown')
		else:
			# duration 1 log inter. | log intervals left
			duration = now-start
			duration = duration/(rerun/cycles) - duration
			print(' Est:', str(duration/60)[0:5], 'mins')

		highest = tally[1]
		lowest = tally[1]
		total = 0
		for i in range(0, 4):
			if players[i].simplex == True:
				continue

			if highest < tally[i]:
				highest = tally[i]
			elif lowest > tally[i]:
				lowest = tally[i]
			total += tally[i]
		deviation = (highest - lowest) / total * 100

		print(' Status;')
		print('  - Iteration       :', rerun, 'of', cycles)
		print('  - Generation      :', generation)
		print(' Game Meta;')
		print('  - Avg. Win        :', str(smoothedWinRate*100)[0:5]+'%')
		print('  - Games Played    :', gameCount)
		print('  - Wins VS Simplex :', winCount)
		print('  - Win Rate        :', str(winRate*100)[0:4]+'%')
		print('  - Win Tally       :', ', '.join([str(val) for val in tally]))
		print('  - Deviation       :', str(deviation)[0:4]+'%')
		print(' Best;')
		print('  - Weights         : [', ', '.join([str(val)[0:5] for val in bestNetwork.linear[0:10]]) + ', ... ]', len(bestNetwork.linear) )
		print('  - Structure       : [', ', '.join([str(val)[0:5] for val in bestNetwork.columns]), ']', len(bestNetwork.columns) )
		print(' Previous;')
		print('  - Weights         : [', ', '.join([str(val)[0:5] for val in prevNetwork.linear[0:10]]) + ', ... ]', len(prevNetwork.linear) )
		print('  - Structure       : [', ', '.join([str(val)[0:5] for val in prevNetwork.columns]), ']', len(prevNetwork.columns)  )
		print(' Progress;')

		file.write(str(now-start)+',')                               # Elapsed
		file.write(str(generation)+',')                              # Generation
		file.write(','.join([str(val) for val in tally])+',')   # Scores[4]
		file.write(str(winRate)+',')                                 # Win Rate
		file.write(str(len(bestNetwork.linear))+',')                 # Number of weights
		file.write(str(len(bestNetwork.columns))+',')                # Number of columns
		file.write(str(max(bestNetwork.columns))+',')                # Largest column
		neurons = 0                                                  # Number of neurons
		for val in bestNetwork.columns:
			neurons += val
		file.write(str(neurons)+'\n')



	for rerun in range(0, cycles):


		# Create mutations of best BOT
		for i in range(0, 4):
			if players[i].simplex == False and players[i].best == False:
				players[i].network = bestNetwork.reproduce(prevNetwork)



		# Make the games best out of 5 to remove chances of fluke
		score = [0, 0, 0, 0]
		for _ in range(0, 5):
			out = game(players)
			games += 1

			winnerScore = out[0]
			winnerIndex = 0
			for k in range(1, 4):
				if out[k] > winnerScore:
					winnerScore = out[k]
					winnerIndex = k
			score[winnerIndex] += 1
			tally[winnerIndex] += 1



		# Determine the best (non-simplex) bot
		greatestScore = 0
		greatestIndex = -1
		for i in range(0, 4):
			if players[i].simplex == True:
				continue

			if score[i] > greatestScore:
				greatestScore = score[i]
				greatestIndex = i

		# Data for logging
		if greatestScore == bestScore:
			print('    - New Varient')
		elif greatestIndex > bestScore:
			print('  - New Best:', score[greatestIndex])
			bestScore = score[greatestIndex]

		# Is this a new best?
		# If so, make it a new breeder
		if players[greatestIndex].best != True:
			generation += 1

			# Mark this player as the best
			# And remove that status from all other players
			for j in range(0, 4):
				if j == i:
					continue
				players[j].best = False
			players[greatestIndex].best = True
			
			prevNetwork = bestNetwork
			bestNetwork = players[greatestIndex].network




		# Add up all wins not including the simplex bot
		for i in range(0, 4):
			if players[i].simplex == True:
				continue

			wins += score[i]




		
		# Remove some data points to help acess
		if rerun % logInterval == 0:

			# Print and write out results
			print('\nStatus:', str(rerun/cycles*100)[0:4] + '%')
			winRate = wins/games
			smoothedWinRate = (smoothedWinRate*3 + winRate) / 4
			now = time.time()
			updateLog(rerun, games, wins)
			wins = 0
			games = 0
			for i in range(0,4):
				tally[i] = 0

	print('Finished\a')
	file.close()

	print('\nEnd Product')
	print(' Best;')
	print('  - Weights:   [', ', '.join([str(val)[0:10] for val in bestNetwork.linear]), ']')
	print('  - Structure: [', ', '.join([str(val)[0:10] for val in bestNetwork.columns]), ']')

	file = open("result.dat", "w+")
	file.write('Result;\n')
	file.write(' - Weights:   [' + ', '.join([str(val) for val in bestNetwork.linear]) + ']\n')
	file.write(' - Structure: [' + ', '.join([str(val) for val in bestNetwork.columns]) + ']\n')
	file.close()





if __name__ == '__main__':
	learnByExample()

File: AI Challenge/code/AI/Cpp/simulate.cpp
#include "./simulate.hpp"

GameResult playRound(Bot* players[4]){
	CardSet deck(true);
	deck.shuffle();


	// Empty player's hands
	for (int i=0; i<4; i++){
		players[i]->reset();
	}

	// Distribute the cards
	int length = deck.size();
	for (int i=0; i<length; i+=4){
		for (int j=0; j<4; j++){
			players[j]->append(deck[i+j]);
		}
	}



	// Round info
	int            i = 0;              // Current player
	int    decompose = 3;              // How many people able to pass in a row
	bool start_round = true;           // Is this the start of the round
	bool     playing = true;          // Is the game finished
	int handSizes[4] = {13,13,13,13};  // Number of cards in each player's hand
	CardSet beat(false);               // The current best play
	CardSet play(false);               // The current play

	// Find which player has
	Card startCard(CardText{'3', 'D'});
	for (int j=0; j<4; j++){
		if (players[j]->has(startCard) == true){
			i = j-1;
			break;
		}
	}

	int passes = 0;

	// Simulate game turn by turn
	while (playing){
		// Detect pass loops
		if (passes > 8){
			std::cerr << "Warn: Pass loop detected, terminating round" << std::endl;
			break;
		}


		// Loop around the group
		i++;
		if (i > 3){
			i = 0;
		}

		// If all players passed
		// Reset the card to beat
		if (decompose == 0){
			beat      = CardSet(false);
			decompose = 3;
		}

		play = players[i]->play(
			start_round,
			beat,
			handSizes
		);
		start_round = false;

		// Handle players passing
		if (play.size() == 0){
			decompose -= 1;
			passes++;
			continue;
		}
		passes = 0;

		// Record the number of cards for each player
		handSizes[i] -= play.size();

		// Check that the play is valid
		if (beat.size() > 1){
			if (play[0].lessThan(&beat[0])){
				std::cerr << "Invalid play: " << play[0].toString() << " < " << beat[0].toString();
			}
		}
		beat = play;

		// This player has played their last card
		if (handSizes[i] < 1){
			playing = false;
			continue;
		}

		// Let all other player's know what happened
		for (int i=0; i<4; i++){
			players[i]->observe(play);
		}
	};



	// Calculate scores
	GameResult res;
	res.winnerIndex = 0;
	res.scores[0] = 0;
	res.scores[1] = 0;
	res.scores[2] = 0;
	res.scores[3] = 0;
	int cards = 0;
	int total = 0;
	for (int i=0; i<4; i++){
		cards = handSizes[i];

		if (cards == 0){
			res.scores[i] = 0;
		}else if (cards < 10){
			res.scores[i] = -1 * cards;
		}else if (cards <= 12) {
			res.scores[i] = -2 * cards;
		}else{
			res.scores[i] = -3 * cards;
		}
		total -= res.scores[i];

		if (res.scores[i] > res.scores[res.winnerIndex]){
			res.winnerIndex = i;
		}
	}
	// Sum of total scores - their score = sum of other players scores
	res.scores[res.winnerIndex] = total + res.scores[res.winnerIndex];


	return res;
};

GameResult playGame(Bot* players[4]){
	GameResult res;
	GameResult rnd;

	res.winnerIndex = 0;
	res.scores[0]   = 0;
	res.scores[1]   = 0;
	res.scores[2]   = 0;
	res.scores[3]   = 0;


	// Play 10 rounds stopping if any score reaches 100
	bool finished = false;
	for (int i=0; i<10; i++){
		if (finished){
			break;
		}

		rnd = playRound(players);
		for (int j=0; j<4; j++){
			res.scores[j] += rnd.scores[j];

			if (res.scores[j] >= 100){
				finished = true;
			}
		}
	}


	// Determine the winner
	int bestIndex = 0;
	for (int i=1; i<4; i++){
		if (res.scores[i] > res.scores[bestIndex]){
			bestIndex = i;
		}
	}
	res.winnerIndex = bestIndex;


	return res;
};





Network train(Network reference){
	// Create bots
	Bot *player[] = {nullptr, nullptr, nullptr, nullptr};
	player[0] = new Bot(true);
	player[1] = new Bot(false);
	player[2] = new Bot(false);
	player[3] = new Bot(false);

	
	// Create breeders
	Network bestNetwork = reference;
	Network prevNetwork = reference;
	int bestIndex       = 0;


	// Training / Log info
	int       games = 0;
	int        wins = 0;
	int      cycles = 5000;
	int logInterval = 50; // 500 long points
	int generations = 0;
	int    tally[4] = {0,0,0,0};
	int   scores[4] = {0,0,0,0};
	float   winRate = 1;
	float  sWinRate = 0;
	int   highScore = 0;
	int    lowScore = 0;
	int  totalScore = 0;
	float deviation = 0;
	float remaining = 0;
	std::string str;


	// CSV file
	std::ofstream csv("log.csv");

	// Duration
	clock_t begin_time = clock();
	float duration;


	// Temp data
	GameResult out;
	int cBest = 0;
	int tick  = 0;


	for (int rerun=0; rerun<cycles; rerun++){

		// Mutate bots
		for (int i=0; i<4; i++){
			if (i == bestIndex || player[i]->simplex == true){
				continue;
			}

			player[i]->network = bestNetwork.reproduce(&prevNetwork);
		}



		// Best of 5 to remove fluke
		scores[0] = 0;
		scores[1] = 0;
		scores[2] = 0;
		scores[3] = 0;
		for (int rnd=0; rnd<5; rnd++){
			out = playGame(player);
			scores[out.winnerIndex]++;
			games += 1;

			for (int i=0; i<4; i++){
				tally[i] += out.scores[i];
			}
			// tally[out.winnerIndex]++;

			if (player[out.winnerIndex]->simplex == false){
				wins++;
			}
		}


		// Determine which player one the most games
		cBest = -1;
		for (int i=0; i<4; i++){
			// Ignore not ANNs
			if (player[i]->simplex == true){
				continue;
			}

			if (cBest == -1){
				cBest = i;
				continue;
			}

			if (scores[i] > scores[cBest]){
				cBest = i;
			}
		}


		// If this is a new champion, give them breeding rights
		if (cBest != bestIndex){
			prevNetwork = bestNetwork;
			bestNetwork = player[cBest]->network;
			bestIndex = cBest;

			generations++;
			std::cout << "   New Generation" << std::endl;
		}


		if (tick >= logInterval){
			tick = 0;

			winRate = float(wins) / games;
			sWinRate = (winRate + sWinRate) / 2;

			duration = float(clock() - begin_time) / CLOCKS_PER_SEC;
			remaining = (duration / rerun) * (cycles - rerun);

			lowScore   = tally[0];
			highScore  = tally[0];
			totalScore = 0;
			for (int i=0; i<4; i++){
				if (lowScore > tally[i]){
					lowScore = tally[i];
				}
				if (highScore < tally[i]){
					highScore = tally[i];
				}

				totalScore += abs(tally[i]);
			}
			deviation = float(abs(highScore - lowScore)) / totalScore;


			str  = "\n\nStatus;\n";
			str += "  - Iteration       : " + std::to_string(rerun) + " of " + std::to_string(cycles) + "\n";
			str += "  - Generation      : " + std::to_string(generations) + "\n";
			str += "  - Duration        : " + std::to_string(duration) + " sec\n";
			str += "  - Estimate Time   : " + std::to_string(remaining) + " mins\n";
			str += "Games;\n";
			str += "  - Avg. Win        : " + std::to_string(sWinRate*100).substr(0,4) + "%\n";
			str += "  - Games Played    : " + std::to_string(games)                           +  "\n";
			str += "  - Wins VS Simplex : " + std::to_string(wins)                            +  "\n";
			str += "  - Win Rate        : " + std::to_string(winRate*100).substr(0,4)         + "%\n";
			
			str += "  - Tally           :";
			for (int i=0; i<4; i++){
				str += " " + std::to_string( tally[i] );
			}
			str += "\n";

			str += "  - Deviation       : " + std::to_string( deviation*100 ).substr(0,4)     + "%\n";

			str += "ANN;\n";
			str += "  - Topology        :";
			int len = bestNetwork.topology.size();
			for (int i=0; i<len; i++){
				str += " "+std::to_string(bestNetwork.topology[i]);
			}
			str += "\n";

			std::cout << str;



			csv << generations << ",";
			csv << winRate << ",";
			csv << deviation << "\n";
			csv.flush();



			tally[0] = 0;
			tally[1] = 0;
			tally[2] = 0;
			tally[3] = 0;
			wins = 0;
			games = 0;

			bestNetwork.save("network.dat");
		}
		tick++;
	}

	std::cout << "\aTranning complete;"<<std::endl;
	std::cout << "  Generations: " << generations << std::endl;
	bestNetwork.save("network.dat");
	csv.close();

	return bestNetwork;
};